### PR TITLE
Initial parts of reputation system

### PR DESCRIPTION
**Pull request by drinkcoffee**
First parts of the reputation system. Uses in-memory storage.

Signed-off-by: Peter  Robinson <drinkcoffee@eml.cc>

### CODE REVIEW

**Comment by drinkcoffee** [comment]
it only uses in-memory objects within the package. I figure that we first need to understand what needs to be stored before we get too far into working out how to persist it. I have however separated out the storing / retrieving of data from the data store from the application logic. Also, given there will be a large set of constants, I have them in a separate file. No tests yet.

File: internal/reputation/reputation_store.go [hunk 1]
```diff
@@ -0,0 +1,33 @@
...
+
+import (
+	"github.com/ConsenSys/fc-retrieval-gateway/pkg/nodeid"
+)
+
+// NOTE: At present all reputation is stored in memory. This will need to change as we move beyond a PoC.
+
+
+func (r *Reputation) getClientReputation(clientNodeID *nodeid.NodeID) (val int, exists bool) {
+	val, exists = r.clients[clientNodeID.ToString()]
```
Comment (wcgcyx) [inline internal/reputation/reputation_store.go:12]: Possibly missing mutex.RLock() ?

**Review by wcgcyx — CHANGES_REQUESTED** [review]
Possibly missing mutex.RLock() when reading from the client reputation map.

File: internal/reputation/reputation_store.go (hunk 1 above)
Comment (drinkcoffee) [reply internal/reputation/reputation_store.go:12]: correct. I will fix this.

Comment (drinkcoffee) [reply internal/reputation/reputation_store.go]: fixed. Added mutex around all reads

File: internal/reputation/reputation_store.go [hunk 2]
```diff
@@ -0,0 +1,38 @@
...
+	clientNodeIDStr := clientNodeID.ToString()
+	var val int
+	var exists bool
+	r.clientsMapLock.Lock()
+	val, exists = r.clients[clientNodeIDStr]
+	if (!exists) {
+		r.clientsMapLock.Unlock()
+		panic("changeClientReputation for non-existant client: " + clientNodeIDStr)
+	}
+	r.clients[clientNodeIDStr] = val + amount
```
Comment (drinkcoffee) [inline internal/reputation/reputation_store.go:36]: Have a two constants: maxClientReputation and minClientReputation. Add check & limit (val + amount) to stay within the min and max reputation.

File: internal/reputation/reputation_store.go [hunk 3]
```diff
@@ -0,0 +1,38 @@
...
+import (
+	"github.com/ConsenSys/fc-retrieval-gateway/pkg/nodeid"
+)
+
+// NOTE: At present all reputation is stored in memory. This will need to change as we move beyond a PoC.
+
+
+func (r *Reputation) getClientReputation(clientNodeID *nodeid.NodeID) (val int, exists bool) {
+	clientNodeIDStr := clientNodeID.ToString()
+	r.clientsMapLock.Lock()
```
Comment (wcgcyx) [inline internal/reputation/reputation_store.go:13]: For read operation, maybe we can use RLock()? map should be safe for concurrent read.

**Pull request merged**